adaptdl/adaptdl/torch/scaling_rules.py

```diff
@@ -287,47 +261,176 @@ def _final_callback(self):
             self._update_avg('sqr_avg', grad_sqr, theta)
             self._update_avg('var_avg', grad_var, theta)
 
-    def step(self, *args, **kwargs):
+    def zero_grad(self, *args, **kwargs):
+        if self._should_zero_grad:
+            self._reset_accumulation()
+        else:
+            warnings.warn("skipping zero_grad for accumulated gradient")
+
+    def patch_optimizer(self, step_fn):
+        """
+        Monkey-patch the optimizer's step function with
+        :meth:`ScalingRuleBase.step`.
+        Arguments:
+            step_fn (function): Scaling rule's step function.
+        """
+        @functools.wraps(self.optimizer.step)
+        def step_wrapper(optim, *args, **kwargs):
+            return step_fn(*args, **kwargs)
+
+        @functools.wraps(self.optimizer.zero_grad)
+        def zero_wrapper(optim, *args, **kwargs):
+            return self.zero_grad(*args, **kwargs)
+        self.optimizer.step = MethodType(step_wrapper, self.optimizer)
+        self.optimizer.zero_grad = MethodType(zero_wrapper, self.optimizer)
+
+    def run_step(self, calculate_lr_factors_fn, *args, **kwargs):
         """
-        Run one optimizer step using Adascale. Essentially just invokes
+        Run one optimizer step. Essentially just invokes
         ``optimizer.step(*args, **kwargs)`` with a scaled learning rate.
 
         Arguments:
-            args: Positional arguments passed to ``optimizer.step``.
-            kwargs: Keyword arguments passed to ``optimizer.step``.
+            calculate_lr_factors_fn (function): Function used to calculate
+            lr factors/multipliers.
         """
-        if not self._adp.require_backward_grad_sync:
-            return
         scale = self._accum_scale * self._accum_count
-
-        initial_lr = [pg["lr"] for pg in self._optimizer.param_groups]
-        for sqr, var, pg in zip(self._state["sqr_avg"], self._state["var_avg"],
-                                self._optimizer.param_groups):
-            sqr, var = max(float(sqr), 0.0), max(float(var), 1e-6)
-            pg["lr"] = (var + sqr) / (var / scale + sqr) * pg["lr"]
+        initial_lr = [pg["lr"] for pg in self.optimizer.param_groups]
+        lr_factors = calculate_lr_factors_fn(scale)
+        for lr_factor, pg in zip(lr_factors, self.optimizer.param_groups):
+            pg["lr"] = lr_factor * pg["lr"]
         self._orig_optimizer_step(*args, **kwargs)
-        for lr, pg in zip(initial_lr, self._optimizer.param_groups):
+        for lr, pg in zip(initial_lr, self.optimizer.param_groups):
             pg["lr"] = lr
-        self._state["progress"] += self.gain(scale)
+        self.state["progress"] += self.gain(scale)
+        self._reset_accumulation()
```

Comment:
Remove this line: it can set the gradients to zero when the user doesn't expect them to be.

Reply:
I might have missed it when I did the rebase. updating it